File: 149.Max_Points_on_a_Line/max_point.cpp
class Solution {
public:
  int maxPoints(vector<Point>& points) {
    if(points.size()<3) return points.size();
    int result=0;
        
    map<pair<int,int>,int> slope_count;
    for(int i=0; i<points.size()-1; i++){
      slope_count.clear();
      int same_point=0;
      int point_max=1;
      for(int j=i+1; j<points.size(); j++){
	pair<int,int> slope;
	if(points[i].x==points[j].x){
	  slope=make_pair(0,1);
	  if(points[i].y==points[j].y){
	    ++same_point; 
	    continue;
	  }
	}
	else{
	  int gcd=GCD(points[j].x-points[i].x,points[j].y-points[i].y);
	  slope=make_pair((points[j].x-points[i].x)/gcd,(points[j].y-points[i].y)/gcd);
	}
	int count=0;
	if(slope_count.find(slope)!=slope_count.end()){
	  count=++slope_count[slope];}
	else{
	  slope_count[slope]=2;
	  count=2;
	}
	if(point_max<count) point_max=count;
      }
      result=max(result,point_max+same_point);
    }
    return result;
  }
private:
  int GCD(int a, int b){
    if(b==0) return a;
    return GCD(b,a%b);
  }
};

File: 149.Max_Points_on_a_Line/max_point.java
/*
 Keywords: HashMap, pair
 Use pair as keys to represent the slope. Map the key to the number of points on the line with this slope.
 */

class Solution {
    class Pair{
        public int first;
        public int second;
        public Pair(int a,int b){
            first=a;
            second=b;
        }
        @Override
	    public int hashCode(){ 
            return this.first*31+this.second;
        }
        @Override
	    public boolean equals(Object p){
            if(p instanceof Pair){
                Pair pp=(Pair) p;
                return (this.first==pp.first) && (this.second==pp.second);
            }
            else return false;
        }
    }
    public int maxPoints(Point[] points) {
        if(points.length<=2) return points.length;
        int res=0;
        HashMap<Pair,Integer> hm=new HashMap<Pair,Integer>();
        for(int i=0;i<points.length-1;i++){
            hm.clear();
            int max_points=1;
            int repeat=0;
            for(int j=i+1;j<points.length;j++){
                Pair p;
                if(points[j].x==points[i].x){
                    p=new Pair(0,1);
                    if(points[j].y==points[i].y){
                        repeat++;
                        continue;
                    }
                }
                else{
                    int gcd=GCD(points[j].x-points[i].x,points[j].y-points[i].y);
                    p=new Pair((points[j].x-points[i].x)/gcd,(points[j].y-points[i].y)/gcd);
                }
                if(!hm.containsKey(p)){
                    hm.put(p,2); 
                } 
                else{
                    hm.put(p,hm.get(p)+1);
                }
                max_points=Math.max(max_points,hm.get(p));
            }
            res=Math.max(res,max_points+repeat);
        }
        return res;
    }
    public int GCD(int a,int b){
        if(b==0) return a;
        else return GCD(b,a%b);
    }
}